### PR TITLE
Support for np.extract

### PR DESCRIPTION
**Pull request by rjenc29**
Support for `np.extract`.  

Initial commit for CI.

### CODE REVIEW

**Comment by rjenc29** [comment]
One of the CI jobs timed out; the others passed.

This PR is to support for `np.extract` and it is relatively trivial.  Let me know what you think when you have a mo, ta!

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -2640,3 +2640,29 @@ def impl(a, dtype=None):
...
+        if a.ndim == 0 and cond.size > 1:
+            raise ValueError(msg)
+
+        if a.ndim == 1 and cond.size > a.size + 1:
+            raise ValueError(msg)
+
+        if a.ndim > 1 and cond.size > a.size:
+            raise ValueError(msg)
+
+        out = [a.flat[idx] for idx in range(min(a.size, cond.size)) if cond.flat[idx]]
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: Line > 80 chars?

File: numba/targets/arraymath.py [hunk 2]
```diff
@@ -2640,3 +2640,29 @@ def impl(a, dtype=None):
...
+
+        if a.size == 0:
+            raise ValueError('Cannot extract from an empty array')
+
+        # the below looks odd but replicates NumPy
+        msg = 'condition shape inconsistent with arr shape'
+        if a.ndim == 0 and cond.size > 1:
+            raise ValueError(msg)
+
+        if a.ndim == 1 and cond.size > a.size + 1:
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: hmmm this is strange... however NumPy does IndexError for:
```
np.extract(np.array([True, False, False, False, True]), np.arange(4))
```
whereas this impl returns `array[0]` ?

File: numba/targets/arraymath.py [hunk 3]
```diff
@@ -2640,3 +2640,29 @@ def impl(a, dtype=None):
...
+    def np_extract_impl(condition, arr):
+        cond = np.asarray(condition)
+        a = np.asarray(arr)
+
+        if a.size == 0:
+            raise ValueError('Cannot extract from an empty array')
+
+        # the below looks odd but replicates NumPy
+        msg = 'condition shape inconsistent with arr shape'
+        if a.ndim == 0 and cond.size > 1:
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: This seems to work under NumPy?
```
In [28]: np.extract(np.array([True, False]), np.array(1))                                                                       
Out[28]: array([1])
```

File: numba/tests/test_np_functions.py [hunk 4]
```diff
@@ -1548,6 +1551,92 @@ def test_ediff1d_exceptions(self):
...
+        for cond in itertools.combinations_with_replacement(element_pool, 4):
+            _check({'condition': cond, 'arr': a})
+            _check({'condition': np.array(cond).reshape(2, 2), 'arr': a})
+
+        a = np.array([1, 2, 3])
+        cond = np.array([])
+        _check({'condition': cond, 'arr': a})
+
+        a = np.array([1, 2, 3])
+        cond = np.array([1, 0, 1, 0])  # but [1, 0, 1, 0, 1] raises
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]: I think I see what's going on... it's whether the position of the last truthy value is in bounds?! `cond = np.array([1, 0, 1, 0, 0])` should also work. I think a runtime check will be needed.

Comment (rjenc29) [reply numba/tests/test_np_functions.py]: Oh boy...

Good spot - I will update the tests and tweak the implementation

File: numba/targets/arraymath.py (hunk 1 above)
Comment (rjenc29) [reply numba/targets/arraymath.py]: Wider monitor?  :)
Done.

File: numba/targets/arraymath.py (hunk 2 above)
Comment (rjenc29) [reply numba/targets/arraymath.py]: Added test

File: numba/targets/arraymath.py (hunk 3 above)
Comment (rjenc29) [reply numba/targets/arraymath.py]: Added test

File: numba/tests/test_np_functions.py (hunk 4 above)
Comment (rjenc29) [reply numba/tests/test_np_functions.py]: Changed the implementation - hopefully this aligns with NumPy better.

File: numba/targets/arraymath.py [hunk 5]
```diff
@@ -2663,3 +2663,27 @@ def impl(a, dtype=None):
...
+
+    def np_extract_impl(condition, arr):
+        cond = np.asarray(condition).flatten()
+        a = np.asarray(arr)
+
+        if a.size == 0:
+            raise ValueError('Cannot extract from an empty array')
+
+        # the following looks odd but replicates NumPy...
+        if bool(cond[-1]) and cond.size > a.size:
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: I'm not sure this check is strong enough, think it needs to walk the `condition` var from index `a.size` to `cond.size` looking for anything truthy. Consider the following...
```python
from numba import njit
import numpy as np

def foo(cond, arr):
    return np.extract(cond, arr)

arr = np.arange(4)
# cond[-1] is `False`, but cond[-2] is OOB
cond = np.array([True, False, True, False, False, True, False])

try:
    foo(cond, arr)
except Exception as e:
    print(e)
else:
    print("no exception raised")

try:
    njit(foo)(cond, arr)
except Exception as e:
    print(e)
else:
    print("no exception raised")
```
gives:
```
index 5 is out of bounds for size 4
no exception raised
```

Comment (rjenc29) [reply numba/targets/arraymath.py]: Oh, sorry - I will add more tests and fix it up.

Comment (rjenc29) [reply numba/targets/arraymath.py]: The last commit should hopefully cover it

**Comment by rjenc29** [comment]
I think the last commit should sort this one out - I will look for / raise a NumPy issue for this as it seems a bit odd.

**Comment by stuartarchibald** [comment]
Thanks for the fixes, this seems to replicate NumPy now, including the slightly unexpected behaviour, good to merge.

**Pull request merged**